### ISSUE TITLE
Unused variable warning IDE0059 triggered on "discard"

### ISSUE THREAD

**Issue by Tragetaschen** [dotnet/roslyn#32923]
**Version Used**: 
16.0.0 Preview 2

**Steps to Reproduce**:
```csharp
private void test()
{
    ReadOnlySequence<byte> a = default;
    var count = 0;
    foreach (var _ in a) // IDE0059 smudge here on _
        ++count;
}
```

**Expected Behavior**:
VS2017 doesn't show that and using `List<byte>` instead of the `ReadOnlySequence<byte>` doesn't either, so I'm not expecting to see this.

**Actual Behavior**:
IDE0059 is triggered and there's a light bulb is VS. However, trying to click the light bulb that appears when hovering the `_` with the mouse just makes it disappear instead of showing possible fixes. The light bulb at the line number just offers to suppress IDE0059 and unrelated code fixes (whitespace for example).

To my knowledge, this foreach loop is the only way to count the number of `ReadOnlyMemory<byte>` that make up the `ReadOnlySequence<byte>`, the `Length` property only counts the total number of bytes in the entire sequence.

**Comment by CyrusNajmabadi** [comment]
Could you clarify what the difference is?

**Comment by Tragetaschen** [comment]
@CyrusNajmabadi 
The code is part of my DbusCore project.
- I have two `ReadOnlySequence<byte>` https://github.com/Tragetaschen/DbusCore/blob/master/src/Dbus/Connection.cs#L129
- And I use all the `ReadOnlyMemory<byte>` in the two sequences to set up the `iovec`s of a `msghdr` for a socket send (https://github.com/Tragetaschen/DbusCore/blob/master/src/Dbus/SocketOperations.cs#L121)

If this is done without copying, AFAIK, I need to count the number of `ReadOnlyMemory<byte>` so I can setup the right number of `iovec`s. That code is also my first take on the new memory management APIs, so I might be missing something fundamentally.

**Comment by mavasani** [comment]
Yes, I mentioned the same question as @CyrusNajmabadi here: https://github.com/dotnet/roslyn/issues/32855#issuecomment-458732054.

_Why is there a loop instead of `count += a.Length`? The diagnostic fires as it seems fishy for a foreach loop to never access the iteration variable._

**Comment by mavasani** [comment]
Regardless, I am going to change the analyzer to not flag on locals named `_`, `_1`, `_2`, etc. as that is an indication of a deliberate suppression/discard. The question about the logic in the repro not using .Length still stands, but is a separate discussion.

**Comment by Tragetaschen** [comment]
@CyrusNajmabadi @mavasani, the sample `ReadOnlySequence<byte>` is the one you are asking for: The Length is the number of bytes in the entire sequence, that is the sum over all bytes in each of the `ReadOnlyMemory<byte>` that back up the sequence.
https://github.com/dotnet/corefx/blob/a7b5f8c08710dedf0baac4706186074ad67b5de1/src/System.Memory/src/System/Buffers/ReadOnlySequence.Helpers.cs#L352

I'm interested in the number of `ReadOnlyMemory<byte>` that are part of the sequence, that is the number of buffers holding the bytes.

**Comment by CyrusNajmabadi** [comment]
@Tragetaschen I'm sorry, as before i'm not understanding.  As before: can you give a small example where .Length produces a different result than counting like you're doing? Thanks!

**Comment by CyrusNajmabadi** [comment]
Thanks @Tragetaschen .  What i was missing was that the enumerator gives you each `ReadOnlyMemory` one at a time.  And you're trying to count those.

Note: you can close this out.  A change @mavasani is making will make it so that we no longer warn if you used `_` as hte name of a temp.

**Comment by Tragetaschen** [comment]
Alright, seems like a good convention

**Comment by mavasani** [comment]
Actually, the change I made was for parameter names, so I need to make another change for special discard local names. I'll keep this bug open to track that change.

**Comment by mavasani** [comment]
I tried your repro on the latest dogfood builds and it does not repro, even for unused loop control variables with regular names:

![image](https://user-images.githubusercontent.com/10605811/52076975-c4b96900-2544-11e9-8076-5aef1d78c72d.png)

I am going to close this bug.

**Comment by Tragetaschen** [comment]
As written in the OP, I haven't been able to repro this for the `List<byte>` (and just checked `byte[]`) instead of `ReadOnlySequence<byte>`. Maybe the `readonly struct`s trigger something extra?

**Comment by mavasani** [comment]
Yes, indeed. Seems like it repros for certain cases only, I found a small standalone repro:

```
public struct S
{
    public Enumerator GetEnumerator() => throw new NotImplementedException();

    public struct Enumerator
    {
        public Enumerator(S sequence) => throw new NotImplementedException();
        public int Current => throw new NotImplementedException();
        public bool MoveNext() => throw new NotImplementedException();
    }
}
class C
{
    void M(S s)
    {
        foreach (var x in s)
        {
        }
    }
}
```
